src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -350,39 +350,93 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
         {
             actualLength = GetTotalDataBytesToWrite();
 
-            int tmpChecksum = WritePosixName(buffer);
+            int tmpChecksum = WriteName(buffer);
             tmpChecksum += WriteCommonFields(buffer, actualLength, TarHelpers.GetCorrectTypeFlagForFormat(TarEntryFormat.Pax, _typeFlag));
             tmpChecksum += WritePosixMagicAndVersion(buffer);
             tmpChecksum += WritePosixAndGnuSharedFields(buffer);
 
             _checksum = WriteChecksum(tmpChecksum, buffer);
         }
 
-        // All formats save in the name byte array only the ASCII bytes that fit.
+        // Gnu and pax save in the name byte array only the UTF8 bytes that fit.
+        // V7 does not support more than 100 bytes so it throws.
         private int WriteName(Span<byte> buffer)
         {
-            ReadOnlySpan<char> src = _name.AsSpan(0, Math.Min(_name.Length, FieldLengths.Name));
-            Span<byte> dest = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
-            int encoded = Encoding.ASCII.GetBytes(src, dest);
-            return Checksum(dest.Slice(0, encoded));
+            ReadOnlySpan<char> name = _name;
+            int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+
+            if (_format is TarEntryFormat.V7 && name.Length != utf16NameTruncatedLength)
+            {
+                throw new ArgumentException(SR.TarEntryNameExceedsMaxLength, ArgNameEntry);
```

Comment:
The semantic here is that `TarWriter.WriteEntry[Async]` throws if `Name` exceeds max length. It could be better to throw on `TarEntry.set_Name`. 
Let me know if you want me to change this behavior.

Reply:
Do we know all the constraints at that point?  If so, I agree that validating the name when it's being set is better, though in practice I don't know what consumption patterns look like and whether it'll actually make a meaningful difference, e.g. if you're just using our higher-level TarFile APIs, the exception's going to come from the same place regardless of where we throw it.

Reply:
> Do we know all the constraints at that point? 

Yes, the only constraints are the entry format and the encoded size of the name. The only downside I see is that it will make this PR a bigger change. 

Regarding the consumption patterns, I assume most people using TarWriter directly want to build a tar archive feeding it with a folder enumeration, and write it down to anything other than a FileStream.